src/foam/core/CSV.js

```diff
@@ -0,0 +1,233 @@
+/**
+ * @license
+ * Copyright 2017 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.csv',
+  name: 'Outputter',
+
+  documentation: 'CSV Outputter.',
+
+  properties: [
+    {
+      class: 'String',
+      name: 'buf_',
+      value: ''
+    },
+    {
+      class: 'String',
+      name: 'delimiter',
+      value: ','
+    },
+    {
+      class: 'String',
+      name: 'nestedObjectSeperator',
+      value: '__'
+    },
+    {
+      class: 'String',
+      name: 'nlStr',
+      value: '\n'
+    },
+    {
+      class: 'String',
+      name: 'undefinedStr',
+      value: ''
+    },
+    {
+      class: 'Boolean',
+      name: 'outputHeaderRow',
+      value: true
+    },
+    {
+      class: 'Function',
+      name: 'propertyPredicate',
+      value: function(o, p) { return ! p.transient; }
+    }
+  ],
+
+  methods: [
+    function out() {
+      for ( var i = 0 ; i < arguments.length ; i++ ) { 
+        this.buf_ += arguments[i];
+      }
+
+      return this;
+    },
+
+    function outputHeader(o) {
+      if ( ! this.outputHeaderRow ) return;
+
+      this.outputPropertyFilteredName(o, '');
+      this.out(this.nlStr);
+    },
+
+    function toCSV(o) {
+      // Outputs object headers
+      this.outputHeader(o);
+      var ret = this.buf_;
+      this.reset(); // reset to avoid retaining garbage
+
+      // Outputs object values
+      this.output(o);
+      this.out(this.nlStr);
+      ret += this.buf_;
+      this.reset();
+
+      return ret;
+    },
+
+    function outputHeaderTitle(o, prefix) {
+      this.out(this.buf_.length ? this.delimiter : '')
+          .out(prefix, o);
+    },
+
+    function outputPropertyName(o, p, prefix) {
+      if ( ! this.propertyPredicate(o, p) ) return;
+
+      // Checks if property is enum, or object with properties of its own
+      if ( p.of && ( ! foam.core.AbstractEnum.isInstance( o[p.name] ) ) ) {
```

Comment:
`p.of` Was used to check if the property was an 'inner object' (with properties of its own). Is there a better/more conventional way to achieve this?

I'm handling Enum's slightly differently to ensure only the ordinal is printed. Otherwise, if treated like a regular object, the ordinal, label and name are all printed. (`JSON.js` prints only the ordinal).